File: server/src/collectors/edinetFilings.js
/**
 * FSA EDINET corporate filings (securities disclosures).
 * https://api.edinet-fsa.go.jp/api/v2/documents.json
 * Requires Subscription-Key header. Without a key we emit a tiny seed batch.
 *
 * Non-spatial — emits one intel item per filing.
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';
import { getEnv } from '../utils/credentials.js';

const SOURCE_ID = 'edinet-filings';
const TIMEOUT_MS = 10000;
const edinetKey = () => getEnv(null, 'EDINET_API_KEY');

function today() {
  return new Date().toISOString().slice(0, 10);
}

export default async function collectEdinetFilings() {
  let items = [];
  let live = false;

  const API_KEY = edinetKey();
  if (API_KEY) {
    try {
      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), TIMEOUT_MS);
      const url = `https://api.edinet-fsa.go.jp/api/v2/documents.json?date=${today()}&type=2&Subscription-Key=${encodeURIComponent(API_KEY)}`;
      const res = await fetch(url, { signal: ctrl.signal });
      clearTimeout(timer);
      if (res.ok) {
        const data = await res.json();
        const list = data?.results ?? [];
        items = list.slice(0, 200).map((d, i) => ({
          uid: intelUid(SOURCE_ID, d.docID, `idx_${i}`),
          title: d.filerName ? `${d.filerName} — ${d.docDescription || d.formCode || 'filing'}` : (d.docDescription || `Filing ${i}`),
          summary: d.docDescription || null,
          author: d.filerName || null,
          language: 'ja',
          published_at: d.submitDateTime ? new Date(d.submitDateTime).toISOString() : null,
          link: d.docID ? `https://disclosure2.edinet-fsa.go.jp/WEEK0010.aspx?docID=${encodeURIComponent(d.docID)}` : null,
          tags: ['filing', d.formCode ? `form:${d.formCode}` : null].filter(Boolean),
          properties: {
            doc_id: d.docID || null,
            form: d.formCode || null,
            doc_type_code: d.docTypeCode || null,
            sec_code: d.secCode || null,
            edinet_code: d.edinetCode || null,
            submit_date: d.submitDateTime || null,
          },
        }));
        live = items.length > 0;
      }
    } catch { /* fall through to seed */ }
  }

  if (items.length === 0) {
    items = [
      { doc: 'seed_edinet_1', filer: 'Toyota Motor', form: '030000', date: today() },
      { doc: 'seed_edinet_2', filer: 'Sony Group',   form: '030000', date: today() },
    ].map((d) => ({
      uid: intelUid(SOURCE_ID, d.doc),
      title: `${d.filer} — seed filing`,
      summary: 'Seed entry (no EDINET API key configured)',
      author: d.filer,
      language: 'ja',
      published_at: new Date(`${d.date}T00:00:00Z`).toISOString(),
      tags: ['filing', 'seed', `form:${d.form}`],
      properties: { doc_id: d.doc, form: d.form, submit_date: d.date, seed: true },
    }));
  }

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items,
    live,
    description: 'FSA EDINET corporate securities disclosures',
    extraMeta: { key_required: true },
  });
}

File: native/collectors/sources/edinet_filings.c
/* collectors/government/sources/edinet_filings.c
 * Port of server/src/collectors/edinetFilings.js (intelEnvelope).
 * FSA EDINET documents.json (Subscription-Key) → one intel row/filing.
 * Gated on EDINET_API_KEY (JS: no key → seed, which we drop → 0 rows).
 * uid = edinet-filings|<d.docID || idx_<i>> (intelUid first non-empty).
 * published_at mirrors d.submitDateTime (source string passthrough; JS does
 * new Date(x).toISOString() — server-tz-coupled, not in uid/props). */
#include "../../source.h"
#include "../../lib/feedlib.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

/* JS truthy string. */
static const char *sv(const cJSON *o, const char *k) {
  const cJSON *v = cJSON_GetObjectItem(o, k);
  return (v && cJSON_IsString(v) && v->valuestring && v->valuestring[0])
           ? v->valuestring : NULL;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *key = getenv("EDINET_API_KEY");
  if (!key || !*key) {
    fprintf(stderr, "[edinet-filings] gated (no EDINET_API_KEY)\n");
    return 0;
  }

  /* today() = new Date().toISOString().slice(0,10) (UTC date) */
  time_t now = time(NULL);
  struct tm g; gmtime_r(&now, &g);
  char day[11];
  snprintf(day, sizeof day, "%04d-%02d-%02d",
           g.tm_year + 1900, g.tm_mon + 1, g.tm_mday);

  char url[512];
  snprintf(url, sizeof url,
    "https://api.edinet-fsa.go.jp/api/v2/documents.json"
    "?date=%s&type=2&Subscription-Key=%s", day, key);

  cJSON *data = feed_get_json(ctx->http, url, 10000);
  if (!data) return -1;

  cJSON *list = cJSON_GetObjectItem(data, "results");
  int n = 0, i = 0;
  if (cJSON_IsArray(list)) {
    cJSON *d;
    cJSON_ArrayForEach(d, list) {
      /* (cap removed: every record of the fetched array is emitted —
       * docs/SOURCE_EXHAUSTIVENESS.md) */
      const char *docID  = sv(d, "docID");
      const char *filer  = sv(d, "filerName");
      const char *ddesc  = sv(d, "docDescription");
      const char *fcode  = sv(d, "formCode");
      const char *dtcode = sv(d, "docTypeCode");
      const char *seccode= sv(d, "secCode");
      const char *edcode = sv(d, "edinetCode");
      const char *subdt  = sv(d, "submitDateTime");

      /* uid: intelUid(SOURCE_ID, d.docID, 'idx_'+i) → first non-empty */
      char idxbuf[24];
      snprintf(idxbuf, sizeof idxbuf, "idx_%d", i);
      const char *rk = docID ? docID : idxbuf;

      /* title = d.filerName
       *   ? `${filerName} — ${docDescription||formCode||'filing'}`
       *   : (docDescription || `Filing ${i}`) */
      char title[512];
      if (filer) {
        const char *tail = ddesc ? ddesc : (fcode ? fcode : "filing");
        snprintf(title, sizeof title, "%s — %s", filer, tail);
      } else if (ddesc) {
        snprintf(title, sizeof title, "%s", ddesc);
      } else {
        snprintf(title, sizeof title, "Filing %d", i);
      }

      /* link = d.docID ? disclosure2 URL : null */
      char *link = NULL;
      if (docID) {
        size_t need = strlen(docID) + 64;
        link = malloc(need);
        snprintf(link, need,
          "https://disclosure2.edinet-fsa.go.jp/WEEK0010.aspx?docID=%s",
          docID);
      }

      /* tags = ['filing', formCode ? 'form:'+formCode : null].filter */
      cJSON *tags = cJSON_CreateArray();
      cJSON_AddItemToArray(tags, cJSON_CreateString("filing"));
      if (fcode) {
        char tb[96];
        snprintf(tb, sizeof tb, "form:%s", fcode);
        cJSON_AddItemToArray(tags, cJSON_CreateString(tb));
      }
      char *tj = cJSON_PrintUnformatted(tags);

      /* properties — EXACT JS key order, each is `X || null` */
      cJSON *p = cJSON_CreateObject();
      cJSON_AddItemToObject(p, "doc_id",
        docID ? cJSON_CreateString(docID) : cJSON_CreateNull());
      cJSON_AddItemToObject(p, "form",
        fcode ? cJSON_CreateString(fcode) : cJSON_CreateNull());
      cJSON_AddItemToObject(p, "doc_type_code",
        dtcode ? cJSON_CreateString(dtcode) : cJSON_CreateNull());
      cJSON_AddItemToObject(p, "sec_code",
        seccode ? cJSON_CreateString(seccode) : cJSON_CreateNull());
      cJSON_AddItemToObject(p, "edinet_code",
        edcode ? cJSON_CreateString(edcode) : cJSON_CreateNull());
      cJSON_AddItemToObject(p, "submit_date",
        subdt ? cJSON_CreateString(subdt) : cJSON_CreateNull());
      char *pj = cJSON_PrintUnformatted(p);

      intel_item it = {0};
      it.remote_key     = rk;
      it.title          = title;
      it.summary        = ddesc;            /* d.docDescription || null */
      it.author         = filer;            /* d.filerName || null */
      it.lang           = "ja";
      it.published_at   = subdt;            /* d.submitDateTime (passthrough) */
      it.link           = link;
      it.record_type    = "edinet-filings";
      it.properties_json = pj;
      it.tags_json      = tj;
      if (sink->emit(sink, &it) >= 0) n++;

      free(pj); free(tj); free(link);
      cJSON_Delete(p); cJSON_Delete(tags);
      i++;
    }
  }
  cJSON_Delete(data);
  fprintf(stderr, "[edinet-filings] emitted %d\n", n);
  return n > 0 ? 0 : -1;
}

static const source_def edinet_filings_def = {
  .id = "edinet-filings", .collector = "government",
  .name = "FSA EDINET Corporate Filings", .name_ja = "EDINET 有価証券報告書",
   .update_interval_sec = 3600, .run = run,
};
REGISTER_SOURCE(edinet_filings_def)
